app/src/main/java/com/techcourse/dao/UserDao.java

```diff
@@ -2,120 +2,56 @@
 
 import com.techcourse.domain.User;
 import com.interface21.jdbc.core.JdbcTemplate;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-import javax.sql.DataSource;
-import java.sql.Connection;
-import java.sql.PreparedStatement;
-import java.sql.ResultSet;
-import java.sql.SQLException;
 import java.util.List;
 
 public class UserDao {
 
-    private static final Logger log = LoggerFactory.getLogger(UserDao.class);
-
-    private final DataSource dataSource;
-
-    public UserDao(final DataSource dataSource) {
-        this.dataSource = dataSource;
-    }
+    private final JdbcTemplate jdbcTemplate;
 
     public UserDao(final JdbcTemplate jdbcTemplate) {
-        this.dataSource = null;
+        this.jdbcTemplate = jdbcTemplate;
     }
 
     public void insert(final User user) {
         final var sql = "insert into users (account, password, email) values (?, ?, ?)";
-
-        Connection conn = null;
-        PreparedStatement pstmt = null;
-        try {
-            conn = dataSource.getConnection();
-            pstmt = conn.prepareStatement(sql);
-
-            log.debug("query : {}", sql);
-
-            pstmt.setString(1, user.getAccount());
-            pstmt.setString(2, user.getPassword());
-            pstmt.setString(3, user.getEmail());
-            pstmt.executeUpdate();
-        } catch (SQLException e) {
-            log.error(e.getMessage(), e);
-            throw new RuntimeException(e);
-        } finally {
-            try {
-                if (pstmt != null) {
-                    pstmt.close();
-                }
-            } catch (SQLException ignored) {}
-
-            try {
-                if (conn != null) {
-                    conn.close();
-                }
-            } catch (SQLException ignored) {}
-        }
+        jdbcTemplate.update(sql, preparedStatement -> {
+            preparedStatement.setString(1, user.getAccount());
+            preparedStatement.setString(2, user.getPassword());
+            preparedStatement.setString(3, user.getEmail());
+        });
```

Comment:
preparedStatement의 인자 세팅을 Dao에서 직접 해주고 있군요~
Dao에서는 `쿼리 작성`, `쿼리에 전달할 인자`에 집중하는 것이 요구사항이었는데요.
Statement 세팅은 JdbcTemplate 내부에서 이루어져도 되지 않을까요?
제가 해석한 요구사항대로라면, 개발자는 밑에와 같이 코드를 작성해야한다고 생각했어요.
제제의 생각은 어떤지 궁금합니다~🤔
```suggestion
        jdbcTemplate.update(sql, user.getAccount(), user.getPassword(), user.getEmail());
```

Reply:
`쿼리에 전달할 인자`에 대해서 1) 인자의 순서, 2) 인자의 타입을 개발자가 직접 입력해서 넘겨주어야 한다고 생각했습니다. 

1) 인자의 순서 : 제안해주신 방법을 적용하면 update 메서드가 `Object... args` 타입을 받게 될텐데요, 이때 for문을 돌면서 순서대로 인자를 셋팅해 줄 수 있을 것 같습니다. 

```java
// 예시
for (int i = 0; i < args.length; i++) {
   statement.setObject(i + 1, args[i]);
}
```
그런데 단순히 이 코드만 보게 된다면 코드의 의미를 쉽게 파악할 수 없을 거라고 생각했습니다. SQL문과 인자의 순서에 따라 값을 셋팅해주는 코드가 함께 있어야 된다고 생각했습니다. 또한 전달받은 순서로 인자를 셋팅하는 것도 실수하기 쉬운 포인트라고 생각했습니다!

2) 인자의 타입 : `Object... args` 타입으로 값을 전달받게 되면 내부에서 분기문으로 set메서드를 처리하거나, setObject로 모든 타입을 뭉뜽그려 처리할 수 있을 것 같아요. 하지만 이 부분은 정확하게 넘겨주는게 안전하다고 생각했습니다.

저도 처음에는 제안해주신 방법으로 적용했으나, 1, 2번에 대해 고민하다보니 함수형 인터페이스를 활용하게 되었습니다.